src/tree_cmd.cpp

```diff
@@ -178,6 +184,139 @@ static void PlaceTree(TileIndex tile, uint32_t r)
 	}
 }
 
+struct BlobHarmonic
+{
+	int amplitude;
+	float phase;
+	int frequency;
+};
+
+struct BlobPosition
+{
+	int x;
+	int y;
+};
+
+BlobPosition operator-(BlobPosition lhs, BlobPosition rhs)
+{
+	int x = lhs.x - rhs.x;
+	int y = lhs.y - rhs.y;
+	return {x, y};
+}
+
+/**
+ * Creates a star-shaped[sic] polygon originating from (0, 0) as defined by the given harmonics.
+ *
+ * @param radius The maximum radius of the polygon. May be smaller, but will not be larger.
+ * @param harmonics a std::vector of the harmonics data.
+ * @param noOfSegments How many segments make up the polygon.
+ */
+std::vector<BlobPosition> createStarShapedPolygon(const int radius, std::vector<BlobHarmonic> harmonics, const int noOfSegments)
+{
+	std::vector<BlobPosition> result;
+
+	float theta = 0;
+	auto step = (M_PI * 2) / noOfSegments; //tau best circle constant
+
+	//divide a circle into a number of equally spaced divisions
+	for(int i = 0; i < noOfSegments; ++i)
+	{
+		float deviation = 0;
+		//add up the values of each harmonic at this segment
+		std::for_each(harmonics.begin(), harmonics.end(), [&deviation, theta](BlobHarmonic &harmonic) {
+			deviation += sin((theta + harmonic.phase) * harmonic.frequency) * harmonic.amplitude;
+		});
+
+		//smooth out changes
+		auto adjustedRadius = (radius / 2.0) + (deviation / 2);
+
+		// add to the final polygon
+		BlobPosition vertex;
+		vertex.x = cos(theta) * adjustedRadius;
+		vertex.y = sin(theta) * adjustedRadius;
+		result.push_back(vertex);
+
+		//proceed to the next segment
+		theta += step;
+	}
+
+	return result;
+}
+
+/**
+ * Creates a random star-shaped[sic] polygon originating from (0, 0).
+ *
+ * @param radius The maximum radius of the blob. May be smaller, but will not be larger.
+ * @param noOfSegments How many segments make up the blob.
+ */
+std::vector<BlobPosition> createRandomStarShapedPolygon(const int radius, const int noOfSegments)
+{
+	std::vector<BlobHarmonic> harmonics;
+
+	// these values are ones i found in my testing that result in suitable- looking polygons that did not self-intersect and fit within a square of radius * radius dimensions.
+
+	harmonics.push_back(BlobHarmonic(radius / 2, Random() / PHASE_DIVISOR, 1));
+	harmonics.push_back(BlobHarmonic(radius / 4, Random() / PHASE_DIVISOR, 2));
+	harmonics.push_back(BlobHarmonic(radius / 8, Random() / PHASE_DIVISOR, 3));
+	harmonics.push_back(BlobHarmonic(radius / 16, Random() / PHASE_DIVISOR, 4));
+
+	return createStarShapedPolygon(radius, harmonics, noOfSegments);
```

Comment:
Pretty sure this can be calculated as part of the loop instead of being stored in a vector.

Reply:
to clarify, do you mean i should remove this function and move the harmonics randomization into the `PlaceTreeGroups` loop?